src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -823,14 +850,15 @@ void OnSubscriptionEstablished(SubscriptionId aSubscriptionId) override
 
     void OnDeallocatePaths(chip::app::ReadPrepareParams && aReadPrepareParams) override {}
 
-    ClusterId mClusterId;
-    uint32_t mValueId;
-    OnSuccessCallbackType mOnSuccess;
+    OnSuccessAttributeCallbackType mOnAttributeSuccess;
+    OnSuccessEventCallbackType mOnEventSuccess;
     OnErrorCallbackType mOnError;
     OnDoneCallbackType mOnDone;
     OnSubscriptionEstablishedCallbackType mOnSubscriptionEstablished;
     app::BufferedReadCallback mBufferedReadAdapter;
     Platform::UniquePtr<app::ReadClient> mReadClient;
+    std::vector<app::AttributePathParams> mAttributePathParamsList;
+    std::vector<app::EventPathParams> mEventPathParamsList;
 };
 
 - (void)readAttributesWithEndpointID:(NSNumber * _Nullable)endpointID
```

Comment:
Here's what I don't understand.  Why can't this function just create a single `MTRAttributeRequestPath` from its arguments and call the new function we are adding, instead of duplicating all the code?

Same for the existing `readEventsWithEndpointID:...`

Reply:
I fully agree with your opinion.
It was thought to minimize modifications to the existing code. I will edit according to your comments.